data_pipeline/calculate_metrics.py

```diff
@@ -7,9 +7,9 @@
 def imputeCols(col):
```

Comment:
Hey Garima, can you write a quick docstring for the method to describe how the imputation is supposed to function?

I ran the following test, but not sure how to interpret the result or if I am just running the test incorrectly:

```
    sample_series = pd.Series([1,2,3,4,5,np.nan, 7,8,9,10,np.nan, 12, 13, 14])   
    new = imputeCols(sample_series)
    print(new)
>>[ 1.  2.  3.  4.  5.  4.  4.  8.  9. 10.  6.  6. 13. 14.]

```

Reply:
Hi Miguel, 

The way the imputation script is working is that it is picking up data from City_Stats (saved in output) and after doing the imputation producing City_Metrics as output

Reason of Imputation- Currently there are lots of missing values in the column. For e.g in tested column, we do not have dailu test figure for few days and then get cumulative test figure the following day. This result in spike in data. Imputation takes this spike data and spread it out to the days when no test figures were available. 

Logic of creating the new column-  We want to see the original numbers in the dashboard without imputation. This implies that there will be few days when the number of test is 0, followed by large number of tests.

Process of creating the new column- I have simply created a new column that is picking up the test value before it is imputed.

Unit Test- I created my dev environment by downloading the raw data and have tested against it. I am getting the results as expected. 

Concerns-  have no idea about what happens after the final data is produced and if it will have any impact on the later processes like dashboards creation.

Does that answer your queries?

Reply:
Hi Garima, I believe Miguel is referring to a quick description of the imputation function that we can include in the documentation.

From what I understand, the code backfills NAs by uniformally redistributing the first valid value of no of tests over the previous null test days.  Eg. If in a week, there were 450 tests attributed to day 7 after 5 days of inactivity, it will assign 64 tests to each of the 7 days.

Additionally, while testing on my own, I came across an issue when correcting -ve values to null when we pass a list but it works when a numpy array or pandas series is passed. In this case, it would be beneficial to state the assumed inputs to the function as well.

Miguel can correct me if this is not the docstring format that we may be following across the project, but as an example we can add the following right at the beginning of the function definition:
```suggestion
def imputeCols(col):
"""
Imputes for null test-days by uniformally redistributing the number of tests observed on the first non-null day observed after these. Eg. 420 tests on day 7 after 6 days of inactivity get distributed across the week as 60 tests/day.

:param col: A numpy array or pandas series containing a daywise record of the number of tests conducted per day.
:returns: A numpy array containing the imputed record of daywise testing numbers.    
"""
```

Reply:
sounds good Harshit! We do not have any defined format yet (Havent come across any doc string as far as I remember). The above format looks good to me. I will update the pull request

Reply:
Thanks, Garima! Yes that helps a lot. and thanks Harshit, for the docstring suggestion